apps/fipsinstall.c

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include "apps.h"
+#include "progs.h"
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/provider.h>
+#include <openssl/params.h>
+
+#define BUFSIZE 4096
+#define DEFAULT_MAC_NAME "HMAC"
+#define DEFAULT_FIPS_SECTION "fips_check_section"
+
+/* Configuration file values */
+#define VERSION_KEY  "version"
+#define VERSION_VAL  "1"
+#define INSTALL_STATUS_VAL "INSTALL_SELF_TEST_KATS_RUN"
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_IN, OPT_CFG, OPT_MACOPT, OPT_VERIFY, OPT_MAC, OPT_SECTION_NAME
+} OPTION_CHOICE;
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    { OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -in fips -cfg "
+      "fips.conf -mac HMAC -macopt digest:SHA256 -macopt hexkey:00"},
+    {"in", OPT_IN, 's', "Input module file name (with no extension)"},
+    {"cfg", OPT_CFG, '>', "config file name to output or verify"},
+    {"mac", OPT_MAC, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm control parameters in n:v form. "
+                                "See 'Supported Controls' in the EVP_MAC_ docs"},
+    {"verify", OPT_VERIFY, '-', "Verify the installed status is correct"},
+    {"section", OPT_SECTION_NAME, 's', "the section name in the config file"},
+
+    {NULL}
```

Comment:
Looking at these parameters, I'm starting to wonder if they shouldn't be restructured a bit.  For example, what is it that this app actually does?  As far as I can see, it writes a config file, or verifies that a given config file is correct.

Looking at other apps, `-in` and `-out` is normally used for the main input / output file, and as already noted, that would be the config file.

So I would suggest restructuring the parameters as follows:

```
    -verify            Verification mode, i.e. verify a config file instead of constructing it
    -in file           Input config file, used when verifying
    -out file          Output config file, used when constructing
    -name str          FIPS provider name
    -module file       FIPS provider module
    -section str       FIPS provider config section name
    -mac str           MAC name
    -macopt str        MAC option in n:v form
```

`-module` and `-section` can be optional, the latter can be constructed from the `-name` value.